hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -978,3 +997,114 @@ def array_for(index_shape, size):
     return result_shape.flatmap(
         lambda index_shape: st.tuples(*(array_for(index_shape, size) for size in shape))
     )
+
+
+def _from_type(thing: Type[Ex]) -> Optional[st.SearchStrategy[Ex]]:
+    """Called by st.from_type to try to infer a strategy for thing using numpy.
+
+    If we can infer a dtype strategy for thing, we return that; otherwise,
+    returns None (or raises).
+    """
+
+    def unpack_generic(thing):
```

Comment:
I'd move these out as top-level functions, with an `_`-prefix - it's a slight performance optimization, but useful as it clarifies that we're not accessing nonlocal variables from the enclosing function.